### PR TITLE
add constraint for rw counter

### PR DESCRIPTION
**Pull request by lightsing**
current design:

1. for all case on word end:
    1. rows[0].rw_counter + 1 == rows[1].rw_counter
    2. rows[0].rwc_inc_left - 1 == rows[1].rwc_inc_left 
2. for non-memory2memory: 
    1. on 0 <= word_index < 31:
        1. rows[0].rw_counter == rows[1].rw_counter
        2. rows[0].rwc_inc_left == rows[1].rwc_inc_left 
    2. on word end:
        1. ~~Memory -> TxLog:~~ (covered by global rule)
            1. ~~rows[0].rw_counter + 1 == rows[1].rw_counter~~
            2. ~~rows[0].rwc_inc_left - 1 == rows[1].rwc_inc_left~~
        2. others:
            1. for read step: rows[0].rw_counter == rows[2].rw_counter
            2. for read step: rows[0].rwc_inc_left == rows[2].rwc_inc_left 
3. for memory2memory: 
    1. on 0 <= word_index < 31:
        1. on read_step: rows[0].rw_counter + 1 == rows[1].rw_counter 
        2. on every step:
            1. rows[0].rw_counter == rows[2].rw_counter
            2. rows[0].rwc_inc_left == rows[2].rwc_inc_left 
    4. ~~on word end:~~ (covered by global rule)
        1. ~~rows[0].rw_counter + 2 == rows[2].rw_counter~~
        2. ~~rows[0].rwc_inc_left - 2 == rows[2].rwc_inc_left~~

### CODE REVIEW

**Comment by lightsing** [comment]
To generialize the constaint and reduce the degree, may use the following approach?
```
rows[0].rw_counter + a == rows[1].rw_counter
rows[0].rwc_inc_left - a == rows[1].rwc_inc_left

rows[0].rw_counter + b == rows[2].rw_counter
rows[0].rwc_inc_left - b == rows[2].rwc_inc_left
```

**Comment by lightsing** [comment]
updated design:

1. on every row:
    1. rows[0].rw_counter + rows[0].rwc_inc_left == rows[1].rw_counter + rows[1].rwc_inc_left
    2. rows[0].rw_counter + diff == rows[1].rw_counter where:
        1. diff = 1 when memory2memory and read step
        2. diff = -1 when memory2memory and write step
        3. diff = 0 otherwise
2. on word end:
    1. rows[0].rw_counter + is_write_step == rows[1].rw_counter

**Comment by lightsing** [comment]
@DreamWuGit test passed with copy_circuit.rs L351-L367 comment out.
https://github.com/scroll-tech/zkevm-circuits/blob/2c00a48a84b866465f67776aafb5133633ec0dbe/zkevm-circuits/src/copy_circuit.rs#L351-L367

**Comment by lightsing** [comment]
Note: some test failed after merge @DreamWuGit

File: zkevm-circuits/src/copy_circuit.rs [hunk 1]
```diff
@@ -324,26 +326,81 @@ impl<F: Field> SubCircuitConfig<F> for CopyCircuitConfig<F> {
                 },
             );
 
-            // addr change: for tx log, addr use addr_slot as index, not increase by 1
+            // for all cases, rw_counter + rwc_inc_left keeps same
             cb.condition(
                 and::expr([
                     not_last_two_rows.expr(),
```
Comment (naure) [inline zkevm-circuits/src/copy_circuit.rs]: Shouldn’t this be `not is_last` (instead of not last two) ?

File: zkevm-circuits/src/copy_circuit.rs [hunk 2]
```diff
@@ -324,26 +326,81 @@ impl<F: Field> SubCircuitConfig<F> for CopyCircuitConfig<F> {
...
+                    not::expr(tag.value_equals(CopyDataType::Padding, Rotation::cur())(meta)),
+                ]),
+                |cb| {
+                    let is_memory2memory = and::expr([
+                        meta.query_advice(is_memory, Rotation::cur()),
+                        meta.query_advice(is_memory, Rotation::next()),
+                    ]);
+                    let diff = select::expr(
+                        is_memory2memory,
+                        select::expr(meta.query_selector(q_step), 1.expr(), -1.expr()),
```
Comment (naure) [inline zkevm-circuits/src/copy_circuit.rs:351]: What are the `1` and `-1` here? It seems that it should be `0` in this case, in the middle of a word.

**Review by naure — COMMENTED** [review]
Questions.

File: zkevm-circuits/src/copy_circuit.rs (hunk 2 above)
Comment (lightsing) [reply zkevm-circuits/src/copy_circuit.rs:351]: it's arranged as a read row, a write row, ...
for memory write it has its own rw counter by increase 1

File: zkevm-circuits/src/copy_circuit.rs (hunk 1 above)
Comment (lightsing) [reply zkevm-circuits/src/copy_circuit.rs]: yeah, you are right.

File: zkevm-circuits/src/copy_circuit.rs (hunk 2 above)
Comment (naure) [reply zkevm-circuits/src/copy_circuit.rs:351]: Still not clear to me, when can the counter go backwards (-1)? 

The most important requirement of the copy circuit is this:
- It looks up an RW operation for each of `[rwc .. rwc + rwc_left]`

Can this be made simpler or clearer how this is enforced?

Comment (lightsing) [reply zkevm-circuits/src/copy_circuit.rs:351]: check this https://docs.google.com/spreadsheets/d/1UyNBdtFM1a6V_amGEc5UhNKLu1eYW2NgXbmp8v-PYvw/edit?usp=sharing

rw counter starts from 0x7c for the first read step and then a write step of 0x7d, so for the next read row, it's need to be same as the fist read row.

constraint using this way is for simplify the condition and reduce the degree.

Comment (DreamWuGit) [reply zkevm-circuits/src/copy_circuit.rs:351]: @naure if you have no more comments, I will merge this PR into memory_opt soon.

**Pull request merged**